mk/external.mk

```diff
@@ -1,41 +1,110 @@
+COMPRESSED_SUFFIX :=
+COMPRESSED_IS_ZIP :=
+COMPRESSED_IS_DIR :=
+EXTRACTOR :=
+VERIFIER :=
+
+# temporarily files to store correct SHA1 value and computed SHA1 value
+# respectively for verification of directory source
+$(eval SHA1_FILE1 := $(shell mktemp))
+$(eval SHA1_FILE2 := $(shell mktemp))
+
+#$(1) : compressed source
+define prologue
+    $(eval _ := $(shell echo "  GET\t$(1)\n"))
+    $(info $(_))
+endef
+
+# $(1), $(2), $(3) : files to be deleted
+define epilogue
+    $(eval _ := $(shell $(RM) $(1) $(2) $(3)))
+endef
+
+# $(1) : compressed source URL
+define download
+    $(eval _ := $(shell curl --progress-bar -O -L -C - "$(strip $(1))"))
+endef
+
+# $(1) : compressed source(.zip or.gz)
+define extract
+    $(eval COMPRESSED_SUFFIX := $(suffix $(1)))
+    $(eval COMPRESSED_IS_ZIP := $(filter $(COMPRESSED_SUFFIX),.zip))
+    $(eval _ := $(if $(COMPRESSED_IS_ZIP),                  \
+		(                                                   \
+			$(eval EXTRACTOR := unzip -d $(OUT) $(1))       \
+		),                                                  \
+		(                                                   \
+			$(eval EXTRACTOR := tar -xf $(1) -C $(OUT))     \
+		)                                                   \
+    ))
+    $(eval _ := $(shell $(EXTRACTOR)))
+endef
+
+# $(1) : correct SHA1 value
+# $(2) : filename or directory path
+# note:
+# 1. for regular file, $(SHA1SUM) command's -c option generates OK for indicating a match
+# 2. for directory, cmp command outputs nothing for indicating a match
+define verify
+    $(eval COMPRESSED_IS_DIR := $(if $(wildcard $(2)/*),1,0))
+    $(eval _ := $(if $(filter 1,$(COMPRESSED_IS_DIR)),             \
+		(                                                          \
```

Comment:
Can you reduce the number of spaces? That is, I would expect more compact indetions.

Reply:
> Can you reduce the number of spaces? That is, I would expect more compact indetions.

Do you mean to not align the backslash?

Reply:
> Do you mean to not align the backslash?

No, I meant something like this:
```
define extract
    $(eval COMPRESSED_SUFFIX := $(suffix $(1)))
    $(eval COMPRESSED_IS_ZIP := $(filter $(COMPRESSED_SUFFIX),.zip))
    $(eval _ := \
        $(if $(COMPRESSED_IS_ZIP), \
            ($(eval EXTRACTOR := unzip -d $(OUT) $(1))), \
            ($(eval EXTRACTOR := tar -xf $(1) -C $(OUT))) \
    ))
    $(eval _ := $(shell $(EXTRACTOR)))
endef
```

Reply:
> > Do you mean to not align the backslash?
> 
> No, I meant something like this:
> 
> ```
> define extract
>     $(eval COMPRESSED_SUFFIX := $(suffix $(1)))
>     $(eval COMPRESSED_IS_ZIP := $(filter $(COMPRESSED_SUFFIX),.zip))
>     $(eval _ := \
>         $(if $(COMPRESSED_IS_ZIP), \
>             ($(eval EXTRACTOR := unzip -d $(OUT) $(1))), \
>             ($(eval EXTRACTOR := tar -xf $(1) -C $(OUT))) \
>     ))
>     $(eval _ := $(shell $(EXTRACTOR)))
> endef
> ```

Got it.